README.md
# trooba-xhr-transport

[![codecov](https://codecov.io/gh/trooba/trooba-xhr-transport/branch/master/graph/badge.svg)](https://codecov.io/gh/trooba/trooba-xhr-transport)
[![Build Status](https://travis-ci.org/trooba/trooba-xhr-transport.svg?branch=master)](https://travis-ci.org/trooba/trooba-xhr-transport) [![NPM](https://img.shields.io/npm/v/trooba.svg)](https://www.npmjs.com/package/trooba)
[![Downloads](https://img.shields.io/npm/dm/trooba.svg)](http://npm-stat.com/charts.html?package=trooba)
[![Known Vulnerabilities](https://snyk.io/test/github/trooba/trooba-xhr-transport/badge.svg)](https://snyk.io/test/github/trooba/trooba-xhr-transport)

XHR transport for trooba pipeline to make CORS RESTful service calls from the browser.

## Get Involved

- **Contributing**: Pull requests are welcome!
    - Read [`CONTRIBUTING.md`](.github/CONTRIBUTING.md) and check out our [bite-sized](https://github.com/trooba/trooba-xhr-transport/issues?q=is%3Aissue+is%3Aopen+label%3Adifficulty%3Abite-sized) and [help-wanted](https://github.com/trooba/trooba-xhr-transport/issues?q=is%3Aissue+is%3Aopen+label%3Astatus%3Ahelp-wanted) issues
    - Submit github issues for any feature enhancements, bugs or documentation problems
- **Support**: Join our [gitter chat](https://gitter.im/trooba) to ask questions to get support from the maintainers and other Trooba developers
    - Questions/comments can also be posted as [github issues](https://github.com/trooba/trooba-xhr-transport/issues)

## Install

```
npm install trooba-xhr-transport --save
```

## Usage

```js
var xhrTransport = require('trooba-xhr-transport');
require('trooba')
    .use(xhrTransport, {
        protocol: 'http:',
        hostname: 'www.google.com'
        socketTimeout: 1000
    })
    .build('client:default')
    .get({
        q: 'nike'
    })
    .set('some', 'header')
    .end(function (err, response) {
        console.log(err, response && response.body)
    });
```

For a real browser example, take a look at [unit test](test/browser.js).
It uses [lasso-js](https://github.com/lasso-js/lasso) and [marko-js](https://github.com/marko-js/marko) to bundle resources into browser page including [trooba xhr implementation](https://github.com/trooba/trooba-xhr-transport/blob/master/test/fixtures/components/app-ajax/index.js#L35). 


CHANGELOG.md
# ChangeLog

## v2.0.1
Fixed: the handler fails to handle 2XX responses
Fixed: Should allow different origin when allow origin is *

## v2.0.0
Upgraded to trooba@2.x, compatible, no release needed

## v2.0.0
Upgraded to trooba@1.x


test/utils.js
'use strict';

var Assert = require('assert');

var Utils = require('..').Utils;

describe(__filename, function () {
    it('should once the function', function () {
        var count = 0;
        function fn() {
            count++;
        }

        var f = Utils.once(fn);
        f();
        f();
        Assert.equal(1, count);
    });

    it('should unslash', function () {
        var path = '/path/to/resource';
        Assert.equal('path/to/resource', Utils.unslash(path));
    });

    it('should mixin', function () {
        var src = {
            foo: 'bar',
            qaz: {
                das: 'rgh'
            },
            arr: [1, 2, 3]
        };
        var result = Utils.mixin(src, {});
        Assert.deepEqual(src, result);

        result = Utils.mixin(src, {
            foo: 'bar1'
        });
        Assert.deepEqual(src, result);

        result = Utils.mixin(src, {
            foo: 'bar1',
            dfg: 'asd'
        });
        Assert.deepEqual({
            foo: 'bar',
            dfg: 'asd',
            qaz: {
                das: 'rgh'
            },
            arr: [1, 2, 3]
        }, result);

        result = Utils.mixin(src, {
            foo: 'bar1',
            dfg: 'asd',
            qaz: {
                das: 'rgh1',
                asd: 'azx'
            }
        });
        Assert.deepEqual({
            foo: 'bar',
            dfg: 'asd',
            qaz: {
                das: 'rgh',
                asd: 'azx'
            },
            arr: [1, 2, 3]
        }, result);

        result = Utils.mixin(src, {
            foo: 'bar1',
            dfg: 'asd',
            qaz: {
                das: 'rgh1',
                asd: 'azx'
            },
            arr: [4]
        });
        Assert.deepEqual({
            foo: 'bar',
            dfg: 'asd',
            qaz: {
                das: 'rgh',
                asd: 'azx'
            },
            arr: [1, 2, 3]
        }, result);

        result = Utils.mixin(src, {
            foo: 'bar1',
            dfg: 'asd',
            qaz: {
                das: 'rgh1',
                asd: 'azx'
            },
            arr: 4
        });
        Assert.deepEqual({
            foo: 'bar',
            dfg: 'asd',
            qaz: {
                das: 'rgh',
                asd: 'azx'
            },
            arr: [1, 2, 3]
        }, result);
    });

    it('should deserializeResponseHeaders', function () {
        var res = {
            headers: 'foo:bar\nqaz:edc\nrfv:tgb'
        };
        Utils.deserializeResponseHeaders(res);
        Assert.deepEqual({
            foo: 'bar',
            qaz: 'edc',
            rfv: 'tgb'
        }, res.headers);

        res = {
            headers: 'foo=dfg'
        };
        Utils.deserializeResponseHeaders(res);
        Assert.deepEqual({}, res.headers);
    });

});


test/invoke.js
'use strict';

var assert = require('assert');
var sinon = require('sinon');
var Service = require('..');

describe(__filename, function() {

    var requestOptions = {
        protocol: 'http:',
        hostname: 'www.ebay.com',
        path: '/jsonpipe'
    };

    describe('verify success scenarios', function() {

        var fakexhr,
            headers = {
                "Content-Type": "application/json",
                "Transfer-Encoding": "chunked"
            };

        before(function() {
            sinon.xhr.supportsXHR = true;
            global.window = {};
            fakexhr = sinon.useFakeXMLHttpRequest();
        });

        after(function() {
            fakexhr.restore();
        });

        it('should process a JSON response with no chunks', function(done) {
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                assert.equal(res.body.id, 7);
                done();
            });

            // increase the chunkSize
            xhr.chunkSize = 20;

            xhr.respond(200, {
                "Content-Type": "application/json"
            }, JSON.stringify({
                "id": 7
            }));
        });

        it('should process 204 response with no content', function(done) {
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err);
                assert.equal(204, res.statusCode);
                assert.equal(undefined, res.body);
                done();
            });

            xhr.respond(204, {
                "Content-Type": "application/json"
            });
        });

        it('should process a JSON response with 1 chunk and ending with \\n\\n', function(done) {
            var count = 1;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 7);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // increase the chunkSize
            xhr.chunkSize = 20;
            xhr.respond(200, headers,
                '{"id": 7}\n\n');
        });

        it('should process a JSON response with 1 chunk and starting with \\n\\n', function(done) {
            var count = 1;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 7);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // increase the chunkSize
            xhr.chunkSize = 20;
            xhr.respond(200, headers,
                '\n\n{"id": 7}');
        });

        it('should process a JSON response with 1 chunk, starting and ending with \\n\\n', function(done) {
            var count = 1;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 7);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // increase the chunkSize
            xhr.chunkSize = 20;
            xhr.respond(200, headers,
                '\n\n{"id": 7}\n\n');
        });

        it('should process a JSON response with 1 chunk, and JSON separated with \\n\\n', function(done) {
            var count = 2;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 2 - count);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // increase the chunkSize
            xhr.chunkSize = 20;
            xhr.respond(200, headers,
                '{"id": 0}\n\n{"id": 1}');
        });

        it('should process a JSON response with multile chunks', function(done) {
            var count = 3;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 3 - count);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // reduce the chunkSize
            xhr.chunkSize = 5;
            xhr.respond(200, headers,
                '{"id": 0}\n\n{"id": 1}\n\n{"id": 2}');
        });

        it('should process a JSON response with multile chunks and bigger buffer size', function(done) {
            var count = 3;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 3 - count);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // reduce the chunkSize
            xhr.chunkSize = 15;
            xhr.respond(200, headers,
                '{"id": 0}\n\n{"id": 1}\n\n{"id": 2}');
        });

        it('should process a JSON response with multile chunks, staring and ending with \\n\\n', function(done) {
            var count = 3;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 3 - count);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // reduce the chunkSize
            xhr.chunkSize = 5;
            xhr.respond(200, headers,
                '\n\n{"id": 0}\n\n{"id": 1}\n\n{"id": 2}\n\n');
        });

        it('should process a JSON response which has Array chunks', function(done) {
            var count = 2;
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.length, 2);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // reduce the chunkSize
            xhr.chunkSize = 5;
            xhr.respond(200, headers,
                '[{"id": 0},{"id": 1}]\n\n[{"id": 2},{"id": 3}]');
        });

        it('should process a multi chunk JSON response separated with the provided option delimiter', function(done) {
            var count = 3;
            var rOptions = Object.create(requestOptions);
            rOptions.delimiter = '\r\r';
            var xhr = Service.invoke(rOptions, function (err, res) {
                assert.ok(!err, err && err.stack);
                if (count > 0) {
                    assert.equal(res.body.id, 3 - count);
                }
                else {
                    assert.equal(res.body, undefined);
                    return done();
                }
                count--;
            });

            // reduce the chunkSize
            xhr.chunkSize = 5;
            xhr.respond(200, headers,
                '\r\r{"id": 0}\r\r{"id": 1}\r\r{"id": 2}\r\r');
        });
    });

    describe('vefiry error scenarios', function() {

        var fakexhr,
            headers = {
                "Content-Type": "application/json",
                "Transfer-Encoding": "chunked"
            };

        before(function() {
            sinon.xhr.supportsXHR = true;
            global.window = {};
            fakexhr = sinon.useFakeXMLHttpRequest();
        });

        after(function() {
            fakexhr.restore();
        });

        it('should call error function on invalid JSON response', function(done) {
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(err);
                assert.equal('parsererror', err.message);
                done();
            });

            xhr.respond(200, headers, '{"id"\n\n}');

        });

        it('should fail once error is encoutered and skipping valid chunk and error on invalid JSON chunk', function(done) {
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(err);
                assert.equal('parsererror', err.message);
                assert.equal('parsererror', err.code);
                assert.equal(500, err.statusCode);
                done();
            });

            xhr.respond(200, headers, '{"id"}\n\n{"id": 1}');

        });

        it('should call error function if timeout exceeded', function(done) {
            var rOptions = Object.create(requestOptions);
            rOptions.socketTimeout = 20;
            var xhr = Service.invoke(rOptions, function (err, res) {
                assert.ok(err);
                assert.equal('ETIMEDOUT', err.code);
                done();
            });

            // Set auto response
            xhr.autoRespond = true;
            xhr.autoRespondAfter = 50;
        });

        it('should call error function on HTTP response code other than 200. e.g. 404', function(done) {
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(err);
                assert.equal('Not Found', err.message);
                assert.equal('Not Found', err.code);
                assert.equal(404, err.statusCode);
                done();
            });

            xhr.respond(404);
        });

        it('should call error function on HTTP response code 400 including service response', function(done) {
            var xhr = Service.invoke(requestOptions, function (err, res) {
                assert.ok(err);
                assert.equal('Bad Request', err.message);
                assert.equal('Bad Request', err.code);
                assert.equal(400, err.statusCode);
                assert.equal('Please provide valid input.', err.response);
                done();
            });
            xhr.respond(400, {}, 'Please provide valid input.');
        });

    });

});


test/browser.js
/* jshint undef: true, unused: true */
'use strict';

var Async = require('async');
var Assert = require('assert');
var Fs = require('fs');
var supertest = require('supertest');
var App = require('./fixtures');
var Shelljs = require('shelljs');
var Browser = require('zombie');
var xhrTransportFactory = require('..');
var Dns = require('dns');
var _lookup = Dns.lookup;

describe(__filename, function() {
    var _cwd;

    afterEach(function() {
        delete global.window;
    });

    before(function (next) {
        this.timeout(120000);

        Dns.lookup = function (host, some, cb) {
            if (host === 'www.test.fake-xyz.com') {
                cb(null, '127.0.0.1', 4);
                return;
            }
            _lookup.apply(Dns, arguments);
        };

        _cwd = process.cwd();
        var root = __dirname + '/fixtures';
        process.chdir(root);

        if (Fs.existsSync(root + '/static')) {
            Shelljs.rm('-rf', root + '/static/*');
        }
        next();
    });

    after(function() {
        process.chdir(_cwd);
        Dns.lookup = _lookup;
    });

    it('should expose service client into browser', function(done) {
        this.timeout(30000);

        App.createTestApp(function (app) {
            // use timeout to allow meta-router to init routes to avoid breaking request context
            // this will also be fixed with latest request-local
            setTimeout(function () {
                supertest(app).get('/page?headers=true').end(function (err, res) {
                    Assert.ok(!err, err && err.stack);
                    console.log(err, res && res.text);
                    var text = res.text;
                    Assert.ok(text.indexOf("'endpoint':{'protocol':'http:','hostname':'www.test.fake-xyz.com','port':7000,'path':'/json','socketTimeout':500}}") !== -1);
                    done();
                });

            }, 2000);

        });
    });

    describe('app flow', function () {
        var app;
        var svr;

        before(function (next) {
            this.timeout(30000);

            Async.series([
                function createServer(next) {
                    App.createTestApp(function (application) {
                        app = application;
                        next();
                    });
                },
                function startServer(next) {
                    svr = app.listen(7000, next);
                }
            ], next);
        });

        after(function () {
            svr && svr.close();
        });

        it('should do full xhr flow in browser', function(done) {
            this.timeout(30000);

            validateXhr(function validate(ctx) {
                Assert.equal('<div id="xhr-response">{"message":"hello world"}</div>',
                    ctx.browser.html('#xhr-response'));
                Assert.equal(1, ctx.countAjaxRequests);
                done();
            });
        });

        it('should do full xhr flow with origin all in browser', function(done) {
            this.timeout(30000);

            validateXhr(function validate(ctx) {
                Assert.equal('<div id="xhr-response">{"message":"hello world"}</div>',
                    ctx.browser.html('#xhr-response'));
                Assert.equal(1, ctx.countAjaxRequests);
                done();
            }, {
                buttonName: 'Do Xhr Origin All Request',
                originHost: 'www.some.other.com'
            });
        });

        it('should handle 204 response with no body from POST request', function(done) {
            this.timeout(30000);

            validateXhr(function validate(ctx) {
                Assert.equal('<div id="xhr-response">204</div>',
                    ctx.browser.html('#xhr-response'));
                Assert.equal(1, ctx.countAjaxRequests);
                done();
            }, {
                buttonName: 'Do Xhr 204 Request',
                originHost: 'www.some.other.com'
            });
        });

        it('should format url', function () {
            var url = xhrTransportFactory.Utils.options2Url({
                method: 'GET',
                search: 'getheaders=true',
                headers: {
                    'x-foo': 'x-bar',
                    'x-qaz': 'x-wsx'
                },
                protocol: 'http:',
                hostname: 'www.test.fake-xyz.com',
                port: 7000,
                path: '/json',
                socketTimeout: 500
            });
            Assert.equal('http://www.test.fake-xyz.com:7000/json?getheaders=true', url);
        });

        it('should do send headers', function(done) {
            this.timeout(30000);

            validateXhr(function validate(ctx) {
                var text = ctx.browser.html('#xhr-response');
                Assert.ok(text.
                    indexOf('"x-foo":"x-bar","x-qaz":"x-wsx"') !== -1);
                Assert.equal(1, ctx.countAjaxRequests);
                done();
            }, {
                buttonName: 'Do Xhr Get Headers'
            });
        });

        it('should do post', function (done) {
            this.timeout(30000);

            validateXhr(function validate(ctx) {
                Assert.equal('<div id="xhr-response">{"foo":"bar"}</div>',
                    ctx.browser.html('#xhr-response'));
                Assert.equal(1, ctx.countAjaxRequests);
                done();
            }, {
                buttonName: 'Do Xhr Post Request'
            });
        });

        it('should do put', function (done) {
            this.timeout(30000);

            validateXhr(function validate(ctx) {
                Assert.equal('<div id="xhr-response">{"foo":"bar"}</div>',
                    ctx.browser.html('#xhr-response'));
                Assert.equal(1, ctx.countAjaxRequests);
                done();
            }, {
                buttonName: 'Do Xhr Put Request'
            });
        });

        it.skip('should do patch', function (done) {
            this.timeout(30000);

            validateXhr(function validate(ctx) {
                Assert.equal('<div id="xhr-response">{"foo":"bar"}</div>',
                    ctx.browser.html('#xhr-response'));
                Assert.equal(1, ctx.countAjaxRequests);
                done();
            }, {
                buttonName: 'Do Xhr Patch Request'
            });
        });

        describe('negative testcases:', function() {
            afterEach(function () {
                delete process.env.TEST_AJAX_NO_RESPONSE;
                delete process.env.TEST_BAD_AJAX_RESPONSE;
                delete process.env.TEST_AJAX_404_RESPONSE;
                delete process.env.TEST_UNKNOWN_HOST;
            });

            it('should handle timeout with xhr', function(done) {
                this.timeout(30000);
                process.env.TEST_AJAX_NO_RESPONSE = 'true';

                validateXhr(function validate(ctx) {
                    Assert.equal('<div id="xhr-response">ERROR: 408</div>',
                        ctx.browser.html('#xhr-response'));
                    Assert.equal(1, ctx.countAjaxRequests);
                    done();
                });

            });

            it('should handle bad response with xhr', function(done) {
                this.timeout(30000);
                process.env.TEST_BAD_AJAX_RESPONSE = 'true';

                validateXhr(function validate(ctx) {
                    Assert.equal('<div id="xhr-response">ERROR: 500</div>',
                        ctx.browser.html('#xhr-response'));
                    Assert.equal(1, ctx.countAjaxRequests);
                    done();
                });
            });

            it('should handle 404 response with xhr', function(done) {
                this.timeout(30000);
                process.env.TEST_AJAX_404_RESPONSE = 'true';

                validateXhr(function validate(ctx) {
                    Assert.equal('<div id="xhr-response">ERROR: 404</div>',
                        ctx.browser.html('#xhr-response'));
                    Assert.equal(1, ctx.countAjaxRequests);
                    done();
                });
            });
        });


    });



});

function validateXhr(validateFn, options) {
    var ctx = {
        countAjaxRequests: 0,
    };
    options = options || {};
    var buttonName = options.buttonName || 'Do Xhr Request';

    Async.series([
        function loadPage(next) {
            var _next = next;
            next = function () {
                _next();
                _next = function noop() {};
            };
            Browser.localhost(options.originHost || 'www.test.fake-xyz.com', 7000);
            var browser = ctx.browser = new Browser();
            browser.debug = true;
            browser.on('xhr', function (event) {
                console.log('xhr:', event);
                if (event === 'loadend') {
                    next();
                }
            });
            browser.on('console', function(level, message) {
                if (/Making Xhr service call.../.test(message) ||
                        /Making Xhr (\w+) service call.../.test(message)) {
                    ctx.countAjaxRequests++;
                }
                else if (message.indexOf('code: \'ETIMEDOUT\'') > 0) {
                    next();
                }
            });
            browser.visit('/page', function startXhr() {
                browser.pressButton(buttonName, function () {});
            });
        },
        function delay(next) {
            setTimeout(next, 500);
        }
    ], function () {
        validateFn(ctx);
    });
}


index.js
/*jslint browser: true*/
'use strict';

var Jsonpipe = require('jsonpipe');
var httpfy = require('trooba-http-api');

/**
 * Ajax/XHR trooba transort for browser side service calls
*/
module.exports = function xhr(pipe, config) {
    pipe.on('request', function onRequest(request) {
        request = Utils.mixin(config,
            request, {});

        invoke(request, function onResponse(err, response) {
            if (err) {
                pipe.throw(err);
                return;
            }

            Utils.deserializeResponseHeaders(response);
            pipe.respond(response);
        });
    });

    httpfy(pipe);
};

function invoke(options, callback) {
    var headers = options.headers ? Object.create(options.headers) : {};
    headers['X-Requested-With'] = 'XMLHttpRequest';
    var err;

    var url = Utils.options2Url(options);
    var _isChunked = false;
    var resHeaders;
    var config = {
        success: function success(data) {
            if (err) {
                return;
            }

            _isChunked = Utils.isChunked(xhr);

            resHeaders = resHeaders || (xhr.getAllResponseHeaders() || '');

            callback(null, {
                statusCode: 200,
                body: data,
                headers: resHeaders
            });
        },
        error: Utils.once(function error(errorMsg) {
            if (err) {
                return;
            }
            if (errorMsg === 'ETIMEDOUT') {
                err = new Error('Connection timeout');
                err.code = 'ETIMEDOUT';
                err.statusCode = 408;
            }
            else {
                err = new Error(errorMsg || 'Resource not found');
                err.code = errorMsg;
                if(errorMsg === 'parsererror') {
                    err.statusCode = 500;
                } else if(errorMsg === 'Bad Request') {
                    err.statusCode = 400;
                } else {
                    err.statusCode = 404;
                }
                err.response = xhr && xhr.response;
            }
        }),
        complete: Utils.once(function complete(statusText) {
            if (err) {
                return callback(err);
            }
            // only do second callback when it is chunked
            if (_isChunked) {
                callback(null, {
                    statusCode: 200,
                    headers: resHeaders,
                    body: undefined
                });
            }
        }),
        timeout: options.socketTimeout, // Number. Set a timeout (in milliseconds) for the request
        method: options.method || 'GET', // String. The type of request to make (e.g. 'POST', 'GET', 'PUT'); default is 'GET'
        headers: headers,
        withCredentials: options.withCredentials,
        data: typeof options.body === 'string' ?
            options.body : options.body !== undefined ?
                httpfy.Utils.stringifyQuery(options.body) : undefined,
        delimiter: options.delimiter
    };

    var xhr = Jsonpipe.flow(url, config);

    // since abort stops any activity and events (in some browsers), we need to simulate
    // the timeout events to complete the flow
    xhr.abort = (function interceptAbort(original) {
        return function abort() {
            // make sure json pipe get correct error code
            err = new Error('Connection timeout');
            err.code = 'ETIMEDOUT';
            err.statusCode = 408;
            original.apply(xhr, arguments);
            config.complete();
        };
    })(xhr.abort);

    return xhr;
}

module.exports.invoke = invoke;

var Utils = {
    isChunked: function isChunked(xhr) {
        var chunked = xhr.getResponseHeader('Transfer-Encoding');
        if (chunked === 'chunked') {
            return true;
        }

        var chromeObj = window.chrome;
        var loadTimes = chromeObj && chromeObj.loadTimes && chromeObj.loadTimes();
        var chromeSpdy = loadTimes && loadTimes.wasFetchedViaSpdy;
        return !!(xhr.getResponseHeader('X-Firefox-Spdy') || chromeSpdy);
    },

    options2Url: function options2Url(options) {
        return (options.protocol || 'http:') + '//' +
            options.hostname +
            (options.port ? (':' + options.port) : '') +
            (options.basepath ? '/' + this.unslash(options.basepath) : '') +
            (options.path ? '/' + this.unslash(options.path) : '') +
            (options.search ?
                '?' + (typeof options.search === 'string' ?
                    options.search :
                    httpfy.Utils.stringifyQuery(options.search)) : '');
    },

    unslash: function unslash(path) {
        return path.replace(/(^\/|\/$)/, '');
    },

    once: function once(fn) {
        var called;
        return function () {
            if (called) {
                return;
            }
            called = true;
            fn.apply(null, arguments);
        };
    },

    mixin: function mixin(src, target) {
        if (src && typeof src === 'object' && target && typeof target === 'object') {
            target = target || {};
            Object.keys(src).forEach(function (key) {
                var srcVal = src[key];
                var targetVal = target[key];
                target[key] = mixin(srcVal, targetVal);
            });
            return target;
        }
        return src || target;
    },

    deserializeResponseHeaders: function deserializeResponseHeaders(response) {
        if (response && response.headers && typeof response.headers === 'string') {
            response.headers = response.headers.split(/\n/)
                .reduce(function reduce(memo, kvp) {
                    var parts = kvp.split(':');
                    if (parts.length > 1) {
                        memo[parts.shift().trim()] = parts.join(':').trim();
                    }
                    return memo;
                }, {});
        }
    }
};

module.exports.Utils = Utils;
